util/jsonpull.c

```diff
@@ -0,0 +1,970 @@
+              g_queue_push_tail (parser->path, path_tail);
+              break;
+            case ']':
+              path_tail = g_queue_pop_tail (parser->path);
+              if (path_tail->parent_type != GVM_JSON_PULL_CONTAINER_ARRAY)
+                {
+                  if (error_message)
+                    *error_message =
+                      g_strdup ("unexpected closing square bracket");
+                  return NULL;
+                }
+              if (path_tail->depth == start_depth)
+                in_expanded_container = FALSE;
+              break;
+            case '}':
+              path_tail = g_queue_pop_tail (parser->path);
+              if (path_tail->parent_type != GVM_JSON_PULL_CONTAINER_OBJECT)
+                {
+                  if (error_message)
+                    *error_message =
+                      g_strdup ("unexpected closing curly brace");
+                  return NULL;
+                }
+              if (path_tail->depth == start_depth)
+                in_expanded_container = FALSE;
+              break;
+            }
+        }
+      gvm_json_pull_parser_next_char (parser);
+    }
+
+  if (parser->last_read_char == GVM_JSON_CHAR_ERROR)
+    {
+      if (error_message)
+        *error_message = gvm_json_read_stream_error_str ();
+      return NULL;
+    }
+  else if (in_expanded_container && parser->last_read_char == GVM_JSON_CHAR_EOF)
+    {
+      if (error_message)
+        *error_message = g_strdup ("unexpected EOF");
+      return NULL;
+    }
+
+  expanded = cJSON_Parse (parser->parse_buffer->str);
+  g_string_truncate (parser->parse_buffer, 0);
+  PARSE_VALUE_NEXT_EXPECT
+
+  if (expanded == NULL && error_message)
+    *error_message = g_strdup ("could not parse expanded container");
+
+  return expanded;
+}
+
+/**
+ * @brief Appends a string path element to a JSONPath string.
+ *
+ * @param[in]  path_elem    The path element to append
+ * @param[in]  path_string  The path string to append to
+ */
+static void
+gvm_json_path_string_add_elem (gvm_json_path_elem_t *path_elem,
+                               GString *path_string)
+{
+  if (path_elem->parent_type == GVM_JSON_PULL_CONTAINER_OBJECT)
+    {
+      gchar *escaped_key = gvm_json_string_escape (path_elem->key, TRUE);
+      g_string_append_printf (path_string, "['%s']", escaped_key);
+      g_free (escaped_key);
+    }
+  else
+    g_string_append_printf (path_string, "[%d]", path_elem->index);
+}
+
+/**
+ * @brief Converts a path as used by a JSON pull parser to a JSONPath string.
+ *
+ * @param[in]  path   The path to convert
+ *
+ * @return Newly allocated string of the path in JSONPath bracket notation
+ */
+gchar *
+gvm_json_path_to_string (GQueue *path)
+{
+  GString *path_string = g_string_new ("$");
+  g_queue_foreach (path, (GFunc) gvm_json_path_string_add_elem, path_string);
+  return g_string_free (path_string, FALSE);
+}
+
+#undef CASE_SPACES
+#undef GVM_JSON_EOF
+#undef GVM_JSON_ERROR
```

Comment:
```suggestion
#undef GVM_JSON_ERROR

```
Trailing newline missing.